gnucash/gnome/assistant-stock-transaction.cpp

```diff
@@ -442,717 +954,456 @@ struct StockTransactionInfo
+        gtk_tree_view_append_column(view, column);
+
+        renderer = gtk_cell_renderer_text_new();
+        gtk_cell_renderer_set_alignment (renderer, 1.0, 0.5);
+        gtk_cell_renderer_set_padding (renderer, 5, 0);
+        column = gtk_tree_view_column_new_with_attributes
+            (_("Units"), renderer,
+             "text", SPLIT_COL_UNITS,
+             "foreground", SPLIT_COL_UNITS_COLOR,
+             nullptr);
+        gtk_tree_view_append_column(view, column);
+
+        return GTK_WIDGET (view);
     }
+};
 
-    auto pinfo = gnc_price_print_info (info->currency, true);
-    gtk_label_set_text (GTK_LABEL (info->price_value), xaccPrintAmount (*price, pinfo));
-}
+static void connect_signals (gpointer, GtkBuilder*);
 
-static void
-refresh_page_cash (GtkWidget *widget, gpointer user_data)
+struct StockAssistantController
 {
-    return;
-}
+    std::unique_ptr<StockAssistantModel> model;
+    std::unique_ptr<StockAssistantView> view;
+    StockAssistantController (GtkWidget *parent, Account* acct)
+        : model (std::make_unique<StockAssistantModel>(acct))
+    {
+        auto builder = gtk_builder_new();
+        gnc_builder_add_from_file (builder, "assistant-stock-transaction.glade",
+                                   "stock_transaction_assistant");
+        this->view = std::make_unique<StockAssistantView>
+            (builder, xaccAccountGetCommodity (acct), this->model->currency, parent);
+        connect_signals (this, builder);
+        g_object_unref (builder);
+        DEBUG ("StockAssistantController constructor\n");
+    };
+    ~StockAssistantController (){ DEBUG ("StockAssistantController destructor\n"); };
+};
 
+/******* implementations ***********************************************/
 static void
-refresh_page_fees (GtkWidget *widget, gpointer user_data)
+stock_assistant_window_destroy_cb (GtkWidget *object, gpointer user_data)
 {
-    auto info = static_cast<StockTransactionInfo*>(user_data);
-    auto capitalize_fees = gtk_toggle_button_get_active
-        (GTK_TOGGLE_BUTTON (info->capitalize_fees_checkbox));
-    gtk_widget_set_sensitive (info->fees_account, !capitalize_fees);
+    auto info = static_cast<StockAssistantController*>(user_data);
+    gnc_unregister_gui_component_by_data (ASSISTANT_STOCK_TRANSACTION_CM_CLASS, info);
+    delete info;
 }
 
 static void
-refresh_page_dividend (GtkWidget *widget, gpointer user_data)
+controller_transaction_type (GtkWidget *widget, StockAssistantController* info)
 {
-    return;
-}
+    if (!info->model->txn_types)
+        return;
 
-static void
-refresh_page_capgains (GtkWidget *widget, gpointer user_data)
-{
-    return;
+    auto type_idx = info->view->get_transaction_type_index();
+    if (type_idx < 0)           // combo isn't initialized yet.
+        return;
+
+    if (!info->model->set_txn_type (type_idx))
+        return;
+
+    info->view->set_txn_type_explanation (info->model->txn_type->explanation);
 }
```

Comment:
The `widget` arg is compulsory when this function is used as a signal callback function...

Reply:
OK. Better naming would help: This is a callback for the transaction_type_combo's "changed" signal, so name it `transaction_type_combo_changed_cb` to clearly identify exactly what this function is for. So would following the `G_CALLBACK` signature, which is `(GtkWidget*, gpointer)`, not `(GtkWidget*, StockAssistantController*)`.  It may save a little typing to silently cast the gpointer, but the convention is to explicitly cast it in the first line of the function. This is actually unsafe unless you can do a runtime check on the cast like the GObject casting macros. In C++ that requires that StockAssistantController is a virtual subclass of something so that you can use dynamic_cast, as in 
```
class Controller; //abstract class
class StockAssistantController : public Controller;

void
random_cb (GtkWidget* widget, gpointer data)
{
    auto info{dynamic_cast<StockAssistantController*>(static_cast<Controller*>(data))};
    g_return_if_fail(info);
//...
}
```
But there are lots of places in GnuCash where we've been too lazy to make a GObject class of something and just done `SomeObject* foo = (SomeObject*)data;`. It only bites us when `data` has been freed and overwritten.

Next, `Info->view->get_transaction_type_index()` is just `return gtk_combo_box_get_active (GTK_COMBO_BOX (transaction_type_combo)); `. That's the same as calling `gtk_combo_box_get_active (GTK_COMBO_BOX (widget))` here, except that the latter is clearer.

Another thing: `info->model->txn_type->explanation` exposes way too much internal  detail so if you decide later that you want to change the structure of info, model, or txn_type you have to change this function too. To avoid that coupling, use functions, e.g. `info->txn_type_explanation(type_idx)`.